### PR TITLE
propose hipe: wallet design

### PR DESCRIPTION
**Pull request by dhh1128**

### CODE REVIEW

File: text/wallets/README.md [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+secrets motivates the following logical organization of identity wallets in Indy:
+
+![wallet composition](composition.png)
+
+The world outside a wallet interfaces with the wallet through a public interface
+provided by indy-sdk, and implemented only once. This is the block labeled
+`encryption, query (wallet core)` in the diagram. The implementation in this layer
+guarantees proper encryption and secret-handling. It also provides some [query
+features](#tags-and-queries).
+Items to be stored in a wallet are referenced by a public handle if they are secrets.
```
Comment (e-nem-y) [inline text/wallets/README.md]: Maybe we should replace `item` with `record` as the indy code is mostly using that term. Not sure it is important, but for the sake of consistency.

File: text/wallets/README.md [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+The way the individual fields are encrypted is shown in the following diagram. Here,
+data is shown as if stored in a relational database with tables. Wallet storage may
+or may not use tables, but regardless of how the storage distributes and divides the
+data, the logical relationships and the encryption shown in the diagram apply.
+
+![encryption](encryption-schema.png)
+
+### Pluggable Storage
+
+Although Indy infrastructure will provide the only one wallet implementation it will allow
```
Comment (e-nem-y) [inline text/wallets/README.md]: Do we need `the` in front of `only one wallet` in this sentence: `Although Indy infrastructure will provide the only one wallet`

File: text/wallets/README.md [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+power outages, and other challenges that would destroy digital data). Of course,
+by leaving the digital realm, the data loses its accessibility over standard APIs.
+
+We anticipate that paper wallets will play a role in backup and recovery, and
+possibly in enabling SSI usage by populations that lack easy access to smartphones
+or the internet. Our wallet design should be friendly to such usage, but physical
+persistence of data is beyond the scope of Indy's plugin storage model and thus
+not explored further in this HIPE. 
+
+### Backup and Recovery
```
Comment (e-nem-y) [inline text/wallets/README.md]: Should we add a bit more detail on how export and import logic as per design https://github.com/hyperledger/indy-sdk/tree/master/doc/design/009-wallet-export-import

Comment (dhh1128) [reply text/wallets/README.md]: Yes, good idea. I've added some details and a link to the internal design doc.

File: text/wallets/README.md (hunk 2 above)
Comment (dhh1128) [reply text/wallets/README.md]: Good catch. Fixed.

File: text/wallets/README.md (hunk 1 above)
Comment (dhh1128) [reply text/wallets/README.md]: Done.

File: text/wallets/README.md [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+* For wallet keys:
+  * Encrypted with master key with random initialization vector.
+
+The 7 "column" keys are concatenated and encrypted with a wallet __master key__,
+ then saved into the metadata of the wallet. This allows the master key to be rotated
+ without re-encrypting all the items in the wallet.
+
+Today, all encryption is done using ChaCha20-Poly1305, with HMAC-SHA256. This is a solid,
+secure encryption algorithm, well tested and widely supported. However, we anticipate the desire
+to use different cipher suites, so in the future we will make the cipher suite pluggaencryption-schema-thumbnail.pnble.
```
[the remark this answers is not in the document]
Comment (dhh1128) [reply text/wallets/README.md]: Good catch. Fixed.

File: text/wallets/README.md (hunk 3 above)
Comment (dhh1128) [reply text/wallets/README.md]: Yes, good suggestion. I've added some detail, plus a link to the design doc.

**Pull request merged**